### DISCUSSION TITLE
"error: initializer element is not constant" when building 5.9.7

### DISCUSSION THREAD

**Discussion by V0174** [strongswan/strongswan#1258] [Q&A] [answered]
We build Strongswan from source since we use a custom directory structure. Everything worked fine with version 5.9.6, but for 5.9.7 I get the following errors when building with GCC:
```
asn1/asn1.c:32:1: error: initializer element is not constant
 const chunk_t ASN1_INTEGER_0 = chunk_from_chars(0x02, 0x01, 0x00);
 ^
asn1/asn1.c:33:1: error: initializer element is not constant
 const chunk_t ASN1_INTEGER_1 = chunk_from_chars(0x02, 0x01, 0x01);
 ^
asn1/asn1.c:34:1: error: initializer element is not constant
 const chunk_t ASN1_INTEGER_2 = chunk_from_chars(0x02, 0x01, 0x02);
 ^
```
The GCC is quite old (4.8.5), is that the reason? I can't easily upgrade it, is there a way to avoid this by setting some configure or make flag?

See the [anonymized log](https://github.com/strongswan/strongswan/files/9450695/strongswan.gcc.issue.txt) for details.

**Comment by tobiasbrunner** [comment, marked as answer]
> The GCC is quite old (4.8.5), is that the reason?

Partially. It's caused with GCC 4.x when compiling with `-std=gnu11`. We don't explicitly enable anything regarding C11 it in our `configure.ac`, but it looks like newer versions of `autoconf` might automatically add such a check via `AC_PROG_CC`:

    checking for gcc option to enable C11 features... -std=gnu11
    checking whether gcc -std=gnu11 understands -c and -o together... yes

Note that with GCC 5.x and newer these chunks are again recognized as constant even with that option passed.

> is there a way to avoid this by setting some configure or make flag?

No. Maybe you can get `configure` to ignore that check (maybe there is a cache variable you could pass as environment variable to make it think there is no such option). Removing `const` in those places might otherwise be the only option.

**Reply by V0174** [reply]
Thanks for your help. Removing the "const"s didn't help, building it with compiled gcc 5.5.0 works.

**Comment by xlapka01** [comment]
For GCC version 4.7.0 the compilation passed when I forced CFLAGS='-std=gnu90' to the configure script.
`CFLAGS='-std=gnu90' ./configure ...`
Moving to a newer compiler is of course preferable, if possible.
